### PR TITLE
[Windows] Pass a function which returns ScopedClipboardInterface instead of ScopedClipboardInterface to PlatformHandlerWin32

### PR DESCRIPTION
**Pull request by moko256**
This PR fixes regression in flutter/flutter#103205 with https://github.com/flutter/flutter/issues/103205#issuecomment-1119733208, and adds test for it.

Fixes flutter/flutter#103205.

No changes in [flutter/tests].

## Pre-launch Checklist

- [X] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [X] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [X] I read and followed the [Flutter Style Guide] and the [C++, Objective-C, Java style guides].
- [X] I listed at least one issue that this PR fixes in the description above.
- [X] I added new tests to check the change I am making or feature I am adding, or Hixie said the PR is test-exempt. See [testing the engine] for instructions on
writing and running engine tests.
- [ ] I updated/added relevant documentation (doc comments with `///`).
- [X] I signed the [CLA].
- [x] All existing and new tests are passing.

If you need help, consider asking for advice on the #hackers-new channel on [Discord].

[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[C++, Objective-C, Java style guides]: https://github.com/flutter/engine/blob/main/CONTRIBUTING.md#style
[testing the engine]: https://github.com/flutter/flutter/wiki/Testing-the-engine
[CLA]: https://cla.developers.google.com/
[flutter/tests]: https://github.com/flutter/tests
[breaking change policy]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
[Discord]: https://github.com/flutter/flutter/wiki/Chat

### CODE REVIEW

File: shell/platform/windows/platform_handler_win32.cc
Comment (moko256) [reply shell/platform/windows/platform_handler_win32.cc]: done

**Comment by tgucio** [comment]
LGTM.
@stuartmorgan, @justinmc - would you be able to take a look so the PR can be merged?

File: shell/platform/windows/platform_handler_win32.h
Comment (stuartmorgan) [inline shell/platform/windows/platform_handler_win32.h:79]: Nit: consider calling this `scoped_clipboard_provider_`, so the name emphasizes that it's returning scoped objects (to make the kind of mistake like the one being fixed here slightly less likely to happen again).

File: shell/platform/windows/platform_handler_win32_unittests.cc
Comment (stuartmorgan) [inline shell/platform/windows/platform_handler_win32_unittests.cc:36]: Apparently I forgot to save a comment; trying again:

Is there a reason you are subclassing instead of using a `friend` test fixture?

Comment (moko256) [reply shell/platform/windows/platform_handler_win32_unittests.cc:36]: I didn't know the feature. I have no reason to avoid it.

Comment (moko256) [reply shell/platform/windows/platform_handler_win32_unittests.cc:36]: The FRIEND_TEST seems to require both target class and tests are in same namespace.
I also tried `friend class testing::PlatformHandlerWin32_ReleaseClipboard_Test;`, but it makes `error: use of undeclared identifier 'testing'`. Are there any macro which exists only on testing?

File: shell/platform/windows/platform_handler_win32.h
Comment (moko256) [reply shell/platform/windows/platform_handler_win32.h:79]: done

Comment (moko256) [reply shell/platform/windows/platform_handler_win32.h]: and I added a comment above this line.

File: shell/platform/windows/platform_handler_win32.h
Comment (justinmc) [inline shell/platform/windows/platform_handler_win32.h:80]: "unnecessary" = "unnecessarily"

Comment (justinmc) [reply shell/platform/windows/platform_handler_win32.h:80]: Nit: Maybe add: "See https://github.com/flutter/flutter/issues/103205."

**Review by justinmc — APPROVED** [review]
Thanks for fixing this and sorry for the breakage!

LGTM but you should also get @stuartmorgan's approval and finish the discussion in https://github.com/flutter/engine/pull/33609#discussion_r882817463.

File: shell/platform/windows/platform_handler_win32.h
Comment (moko256) [reply shell/platform/windows/platform_handler_win32.h]: done

File: shell/platform/windows/platform_handler_win32_unittests.cc
Comment (cbracken) [reply shell/platform/windows/platform_handler_win32_unittests.cc:36]: I don't know that we've ever explicitly documented _not_ to do this in the engine, but I'll be honest; I had no idea `FRIEND_TEST` existed as part of googletest.

The namespacing issue is unfortunate. I wonder if we should just fix this across the board in our tests. A `testing` namespace is certainly useful for utility code to be used in tests, but not sure why we've elected to place tests themselves in it. I'll file a separate issue for discussion if we don't have one already.

**Review by cbracken — APPROVED** [review]

**Comment by tgucio** [comment]
Good to go, if someone could slap the "waiting for tree to go green" label that'd be great.
Thanks @moko256 for the fix.

**Pull request merged**
